include/history_buffer.hpp

```diff
@@ -0,0 +1,224 @@
+ * It is optimised for single-threaded use and does not provide thread-safety or multi-writer/multi-reader support.
+ *
+ * Example usage:
+ * gr::history_buffer<int, 5> buffer;
+ * buffer.push_back(1);  // buffer: [1] (size: 1, capacity: 5)
+ * buffer.push_back(2);  // buffer: [1, 2] (size: 2, capacity: 5)
+ * buffer.push_back(3);  // buffer: [1, 2, 3] (size: 3, capacity: 5)
+ * buffer.push_back(4);  // buffer: [1, 2, 3, 4] (size: 4, capacity: 5)
+ * buffer.push_back(5);  // buffer: [1, 2, 3, 4, 5] (size: 5, capacity: 5)
+ * buffer.push_back(6);  // buffer: [2, 3, 4, 5, 6] (size: 5, capacity: 5)
+ *
+ * // :
+ * buffer[0];     // value: 6 - unchecked access of last/actual sample
+ * buffer[-1];    // value: 5 - unchecked access of previous sample
+ * ...
+ * buffer.at(0);  // value: 6 - checked access of last/actual sample
+ * buffer.at(-1); // value: 5 - checked access of last/actual sample
+ * ...
+ * buffer.get_span(0, 3);  // span: [4, 5, 6]
+ * buffer.get_span(1, 3);  // span: [3, 4, 5]
+ * buffer.get_span(0);     // span: [2, 3, 4, 5, 6]
+ * buffer.get_span(1);     // span: [2, 3, 4, 5]
+ */
+template<typename T, std::size_t N = std::dynamic_extent, typename Allocator = std::allocator<T>>
+class history_buffer {
+    using signed_index_type = std::make_signed_t<std::size_t>;
+    using buffer_type       = typename std::conditional_t<N == std::dynamic_extent, std::vector<T, Allocator>, std::array<T, N * 2>>;
+    using size_type   = typename std::conditional_t<N == std::dynamic_extent, std::size_t, std::size_t>;
+
+    buffer_type _buffer;
+    std::size_t _capacity = N;
+    std::size_t _write_position{ 0 };
+    std::size_t _size{ 0 };
+
+    /**
+     * @brief maps the logical index to the physical index in the buffer.
+     */
+    constexpr inline std::size_t
+    map_index(signed_index_type index) const noexcept {
+        if constexpr (N == std::dynamic_extent) {
+            return (_write_position + _capacity - 1 + index) % _capacity;
+        } else {
+            return (_write_position + N - 1 + index) % N;
+        }
+    }
+
+public:
+    constexpr explicit history_buffer() noexcept { static_assert(N != std::dynamic_extent, "need to specify capacity"); }
+
+    constexpr explicit history_buffer(size_t capacity) : _buffer(capacity * 2), _capacity(capacity) {
+        if (capacity == 0) {
+            throw std::out_of_range("capacity is zero");
+        }
+        static_assert(N == std::dynamic_extent, "incompatible fixed capacity and using capacity argument");
+    }
+
+    /**
+     * @brief Adds an element to the end expiring the oldest element beyond the buffer's capacities.
+     */
+    constexpr void
+    push_back(const T &value) noexcept {
+        _buffer[_write_position]             = value;
+        _buffer[_write_position + _capacity] = _buffer[_write_position];
+        ++_write_position;
+        [[unlikely]] if (_write_position == _capacity) { _write_position = 0; }
```

Comment:
This compiles? `if (_write_position == _capacity) [[unlikely]] {` is the attribute placement I know.